a2s/byteio.py

```diff
@@ -1,80 +1,121 @@
+]
+
 
+class ByteReader:
+    __slots__ = (
+        "stream",
+        "endian",
+        "encoding",
+    )
 
-class ByteReader():
-    def __init__(self, stream, endian="=", encoding=None):
-        self.stream = stream
-        self.endian = endian
-        self.encoding = encoding
+    def __init__(self, stream: io.BytesIO, endian: str = "=", encoding: Optional[str] = None) -> None:
+        self.stream: io.BytesIO = stream
+        self.endian: str = endian
+        self.encoding: Optional[str] = encoding
 
-    def read(self, size=-1):
+    def read(self, size: int = -1) -> bytes:
         data = self.stream.read(size)
         if size > -1 and len(data) != size:
             raise BufferExhaustedError()
 
         return data
 
-    def peek(self, size=-1):
+    def peek(self, size: int = -1) -> bytes:
         cur_pos = self.stream.tell()
         data = self.stream.read(size)
         self.stream.seek(cur_pos, io.SEEK_SET)
         return data
 
-    def unpack(self, fmt):
-        fmt = self.endian + fmt
+    def unpack(self, fmt: STRUCT_OPTIONS) -> Tuple[Any, ...]:
+        new_fmt = self.endian + fmt
         fmt_size = struct.calcsize(fmt)
-        return struct.unpack(fmt, self.read(fmt_size))
+        return struct.unpack(new_fmt, self.read(fmt_size))
 
-    def unpack_one(self, fmt):
+    def unpack_one(self, fmt: STRUCT_OPTIONS) -> Any:
         values = self.unpack(fmt)
         assert len(values) == 1
         return values[0]
 
-    def read_int8(self):
+    def read_int8(self) -> int:
         return self.unpack_one("b")
 
-    def read_uint8(self):
+    def read_uint8(self) -> int:
         return self.unpack_one("B")
 
-    def read_int16(self):
+    def read_int16(self) -> int:
         return self.unpack_one("h")
 
-    def read_uint16(self):
+    def read_uint16(self) -> int:
         return self.unpack_one("H")
 
-    def read_int32(self):
+    def read_int32(self) -> int:
         return self.unpack_one("l")
 
-    def read_uint32(self):
+    def read_uint32(self) -> int:
         return self.unpack_one("L")
 
-    def read_int64(self):
+    def read_int64(self) -> int:
         return self.unpack_one("q")
 
-    def read_uint64(self):
+    def read_uint64(self) -> int:
         return self.unpack_one("Q")
 
-    def read_float(self):
+    def read_float(self) -> float:
         return self.unpack_one("f")
 
-    def read_double(self):
+    def read_double(self) -> float:
         return self.unpack_one("d")
 
-    def read_bool(self):
+    def read_bool(self) -> bool:
         return bool(self.unpack_one("b"))
 
-    def read_char(self):
+    def read_char(self) -> str:
         char = self.unpack_one("c")
         if self.encoding is not None:
             return char.decode(self.encoding, errors="replace")
         else:
-            return char
+            return char.decode(DEFAULT_ENCODING, errors="replace")
```

Comment:
Both this method and `read_cstring` use the return type as strings, regardless of if they are `bytes` instances or not. As such I have added this. If this is in error please let me know.

Reply:
Not decoding the strings and keeping them as bytes was an afterthought, but intentional, so typing them as str was lazy on my part. This is not right and will break code depending on undecoded strings. The simple fix would be to make a type alias for `str` or `bytes` and use that in all places, but that kind of defeats the purpose of typing. Maybe you can come up with something better.

Reply:
I currently have them as `Union[str, bytes]` and will leave it to user code to handle them as needed. :-)